Error instantiating QModelMenu with PySide6 6.5.0

* app-model version: `main` (`0.1.2`)
* Python version: `3.10.10`
* Operating System: macOS 13.3

### Description
There is a problem instantiating the `QModelMenu` with PySide 6.5.0. It looks like multiple inheritance gets a bit trickier here as now QWidget is effectively calling `super().__init__`? This feels like a pretty big change so perhaps it should be reported on the Qt bug tracker.

Try this demo script to see what I mean:

```python
# qwidget_demo.py
from PySide6.QtCore import QObject
from PySide6.QtWidgets import QApplication, QWidget

app = QApplication([])


class A(QObject):
    def __init__(self, *args, **kwargs):
        print("A.__init__", args, kwargs)


class B(QWidget, A):
    def __init__(self, *args, **kwargs):
        QWidget.__init__(self, *args, **kwargs)
        print("B.__init__", args, kwargs)


print("instantiating A")
A()

print("instantiating B")
B()
```

#### output with PySide 6.4.2
```
❯ python qwidget_demo.py 
instantiating A
A.__init__ () {}
instantiating B
B.__init__ () {}
```
#### output with PySide 6.5.0
```
❯ python qwidget_demo.py 
instantiating A
A.__init__ () {}
instantiating B
A.__init__ () {}
B.__init__ () {}
```

### What I Did
Create a virtual environment with the newly released PySide6 v6.5.0:
```
% python -m venv venv
% source venv/bin/activate
% pip install pyside6==6.5.0
% pip install -e '.[qt,test,test-qt]'
% pytest
```

Tests are failing for `QMenu` with this rather confusing output:

```
self = <[RuntimeError("'__init__' method of object's base class (QModelMenu) not called.") raised in repr()] QModelMenu object at 0x12e46ae00>, menu_id = 'help'
app = Application('complete_test_app'), title = None, parent = None

    def __init__(
        self,
        menu_id: str,
        app: Union[str, Application],
        title: Optional[str] = None,
        parent: Optional[QWidget] = None,
    ):
>       QMenu.__init__(self, parent)
E       TypeError: _MenuMixin.__init__() missing 1 required positional argument: 'app'

src/app_model/backends/qt/_qmenu.py:144: TypeError
```

### Possible solution?

I'm not sure what the best solution is here - I don't have a ton of experience with multiple inheritance or mixins. One thing that seems to work is to give `_MenuMixin` a dummy `__init__` like this, but it doesn't necessarily feel like the right way to do it:

```diff
--- a/src/app_model/backends/qt/_qmenu.py
+++ b/src/app_model/backends/qt/_qmenu.py
@@ -44,7 +44,10 @@ class _MenuMixin(QObject):
     _app: Application
     _menu_id: str
 
-    def __init__(
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def _init_mixin(
         self,
         menu_id: str,
         app: Union[str, Application],
@@ -141,8 +144,9 @@ class QModelMenu(QMenu, _MenuMixin):
         title: Optional[str] = None,
         parent: Optional[QWidget] = None,
     ):
+        print(QModelMenu.__mro__)
         QMenu.__init__(self, parent)
-        _MenuMixin.__init__(self, menu_id, app)
+        self._init_mixin(menu_id, app)
         if title is not None:
             self.setTitle(title)
         self.aboutToShow.connect(self._on_about_to_show)
@@ -217,7 +221,7 @@ class QModelToolBar(QToolBar, _MenuMixin):
     ) -> None:
         self._exclude = exclude
         QToolBar.__init__(self, parent)
-        _MenuMixin.__init__(self, menu_id, app)
+        self._init_mixin(menu_id, app)
         if title is not None:
             self.setWindowTitle(title)
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -81,7 +81,7 @@ jobs:
           python -m pip install ${{ matrix.qt-backend }}
 
       - name: Test
-        uses: coactions/setup-xvfb@v1
+        uses: aganders3/headless-gui@v1
         with:
           run: python -m pytest -s --cov=app_model --cov-report=xml --cov-report=term-missing --color=yes
 

--- a/.github/workflows/cron.yml
+++ b/.github/workflows/cron.yml
@@ -41,7 +41,7 @@ jobs:
           python -m pip install --pre ${{ matrix.qt-backend }}
 
       - name: Test
-        uses: GabrielBB/xvfb-action@v1
+        uses: aganders3/headless-gui@v1
         with:
           run: python -m pytest -s --color=yes
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -85,11 +85,11 @@ line-length = 88
 src = ["src", "tests"]
 target-version = "py38"
 extend-select = [
-    "E",    # style errors
-    "F",    # flakes
-    "D",    # pydocstyle
+    "E", # style errors
+    "F", # flakes
+    "D", # pydocstyle
     "I", # isort
-    "U",    # pyupgrade
+    "U", # pyupgrade
     # "N",  # pep8-naming
     # "S",  # bandit
     "C",    # flake8-comprehensions
@@ -123,7 +123,10 @@ keep-runtime-typing = true
 # https://docs.pytest.org/en/6.2.x/customize.html
 [tool.pytest.ini_options]
 minversion = "6.0"
-filterwarnings = ["error", "ignore:QPixmapCache.find:DeprecationWarning:"]
+filterwarnings = [
+    "error",
+    "ignore:Enum value:DeprecationWarning:superqt",
+]
 
 # https://mypy.readthedocs.io/en/stable/config_file.html
 [tool.mypy]

--- a/src/app_model/backends/qt/__init__.py
+++ b/src/app_model/backends/qt/__init__.py
@@ -9,7 +9,7 @@ from ._qkeymap import (
     qmods2modelmods,
 )
 from ._qmainwindow import QModelMainWindow
-from ._qmenu import QModelMenu, QModelMenuBar, QModelSubmenu
+from ._qmenu import QModelMenu, QModelMenuBar, QModelSubmenu, QModelToolBar
 from ._util import to_qicon
 
 __all__ = [
@@ -25,6 +25,7 @@ __all__ = [
     "QModelMenu",
     "QModelMenuBar",
     "QModelSubmenu",
+    "QModelToolBar",
     "qmods2modelmods",
     "to_qicon",
 ]

--- a/src/app_model/backends/qt/_qmenu.py
+++ b/src/app_model/backends/qt/_qmenu.py
@@ -3,15 +3,15 @@ from __future__ import annotations
 from typing import (
     TYPE_CHECKING,
     Collection,
+    Iterable,
     List,
     Mapping,
     Optional,
-    Protocol,
     Set,
     Union,
+    cast,
 )
 
-from qtpy.QtCore import QObject
 from qtpy.QtWidgets import QMenu, QMenuBar, QToolBar
 
 from app_model import Application
@@ -29,99 +29,67 @@ if TYPE_CHECKING:
     from qtpy.QtWidgets import QAction, QWidget
 
 
-# fmt: off
-class _AcceptsMenus(Protocol):
-    _app: Application
-    def clear(self) -> None: ...
-    def addMenu(self, menu: QMenu) -> None: ...
-    def addAction(self, menu: QAction) -> None: ...
-    def addSeparator(self) -> None: ...
+def _rebuild(
+    menu: QMenu | QToolBar,
+    app: Application,
+    menu_id: str,
+    include_submenus: bool = True,
+    exclude: Optional[Collection[str]] = None,
+) -> None:
+    """Rebuild menu by looking up self._menu_id in menu_registry."""
+    menu.clear()
+    _exclude = exclude or set()
 
-# fmt: on
-
-
-class _MenuMixin(QObject):
-    _app: Application
-    _menu_id: str
-
-    def __init__(
-        self,
-        menu_id: str,
-        app: Union[str, Application],
-    ):
-        assert isinstance(menu_id, str), f"Expected str, got {type(menu_id)!r}"
-        self._menu_id = menu_id
-        self._app = Application.get_or_create(app) if isinstance(app, str) else app
-        self.setObjectName(menu_id)
-        self.rebuild()
-        self._app.menus.menus_changed.connect(self._on_registry_changed)
-        self.destroyed.connect(self._disconnect)
-
-    def _disconnect(self) -> None:
-        self._app.menus.menus_changed.disconnect(self._on_registry_changed)
-
-    def _on_registry_changed(self, changed_ids: Set[str]) -> None:
-        if self._menu_id in changed_ids:
-            self.rebuild()
-
-    def rebuild(
-        self: _MenuMixin,
-        include_submenus: bool = True,
-        exclude: Optional[Collection[str]] = None,
-    ) -> None:
-        """Rebuild menu by looking up self._menu_id in menu_registry."""
-        self.clear()
-        _exclude = exclude or set()
-
-        groups = list(self._app.menus.iter_menu_groups(self._menu_id))
-        n_groups = len(groups)
-        for n, group in enumerate(groups):
-            for item in group:
-                if isinstance(item, SubmenuItem) and include_submenus:
-                    submenu = QModelSubmenu(item, self._app, parent=self)
-                    self.addMenu(submenu)
-                elif item.command.id not in _exclude:  # type: ignore
-                    action = QMenuItemAction(
-                        item, app=self._app, parent=self  # type: ignore
-                    )
-                    self.addAction(action)
-            if n < n_groups - 1:
-                self.addSeparator()
-
-    def update_from_context(
-        self, ctx: Mapping[str, object], _recurse: bool = True
-    ) -> None:
-        """Update the enabled/visible state of each menu item with `ctx`.
-
-        See `app_model.expressions` for details on expressions.
-
-        Parameters
-        ----------
-        ctx : Mapping
-            A namespace that will be used to `eval()` the `'enablement'` and
-            `'when'` expressions provided for each action in the menu.
-            *ALL variables used in these expressions must either be present in
-            the `ctx` dict, or be builtins*.
-        _recurse : bool
-            recursion check, internal use only
-        """
-        for action in self.actions():
-            if isinstance(action, QMenuItemAction):
-                action.update_from_context(ctx)
-            elif not QT6 and isinstance(menu := action.menu(), QModelMenu):
-                menu.update_from_context(ctx)
-            elif isinstance(parent := action.parent(), QModelMenu):
-                # FIXME: this is a hack for Qt6 that I don't entirely understand.
-                # QAction has lost the `.menu()` method, and it's a bit hard to find
-                # how to get to the parent menu now. Checking parent() seems to work,
-                # but I'm not sure if it's the right thing to do, and it leads to a
-                # recursion error.  I stop it with the _recurse flag here, but I wonder
-                # whether that will cause other problems.
-                if _recurse:
-                    parent.update_from_context(ctx, _recurse=False)
+    groups = list(app.menus.iter_menu_groups(menu_id))
+    n_groups = len(groups)
+    for n, group in enumerate(groups):
+        for item in group:
+            if isinstance(item, SubmenuItem) and include_submenus:
+                submenu = QModelSubmenu(item, app, parent=menu)
+                cast("QMenu", menu).addMenu(submenu)
+            elif item.command.id not in _exclude:  # type: ignore
+                action = QMenuItemAction(item, app=app, parent=menu)  # type: ignore
+                menu.addAction(action)
+        if n < n_groups - 1:
+            menu.addSeparator()
 
 
-class QModelMenu(QMenu, _MenuMixin):
+def _update_from_context(
+    actions: Iterable[QAction], ctx: Mapping[str, object], _recurse: bool = True
+) -> None:
+    """Update the enabled/visible state of each menu item with `ctx`.
+
+    See `app_model.expressions` for details on expressions.
+
+    Parameters
+    ----------
+    actions : Iterable[QAction]
+        Actions to update.
+    ctx : Mapping
+        A namespace that will be used to `eval()` the `'enablement'` and
+        `'when'` expressions provided for each action in the menu.
+        *ALL variables used in these expressions must either be present in
+        the `ctx` dict, or be builtins*.
+    _recurse : bool
+        recursion check, internal use only
+    """
+    for action in actions:
+        if isinstance(action, QMenuItemAction):
+            action.update_from_context(ctx)
+        elif not QT6 and isinstance(menu := action.menu(), QModelMenu):
+            menu.update_from_context(ctx)
+        elif isinstance(parent := action.parent(), QModelMenu):
+            # FIXME: this is a hack for Qt6 that I don't entirely understand.
+            # QAction has lost the `.menu()` method, and it's a bit hard to find
+            # how to get to the parent menu now. Checking parent() seems to work,
+            # but I'm not sure if it's the right thing to do, and it leads to a
+            # recursion error.  I stop it with the _recurse flag here, but I wonder
+            # whether that will cause other problems.
+            if _recurse:
+                parent.update_from_context(ctx, _recurse=False)
+
+
+class QModelMenu(QMenu):
     """QMenu for a menu_id in an `app_model` MenusRegistry.
 
     Parameters
@@ -142,7 +110,17 @@ class QModelMenu(QMenu, _MenuMixin):
         parent: Optional[QWidget] = None,
     ):
         QMenu.__init__(self, parent)
-        _MenuMixin.__init__(self, menu_id, app)
+
+        # NOTE: code duplication with QModelToolBar, but Qt mixins and multiple
+        # inheritance are problematic for some versions of Qt, and for typing
+        assert isinstance(menu_id, str), f"Expected str, got {type(menu_id)!r}"
+        self._menu_id = menu_id
+        self._app = Application.get_or_create(app) if isinstance(app, str) else app
+        self.setObjectName(menu_id)
+        self.rebuild()
+        self._app.menus.menus_changed.connect(self._on_registry_changed)
+        self.destroyed.connect(self._disconnect)
+
         if title is not None:
             self.setTitle(title)
         self.aboutToShow.connect(self._on_about_to_show)
@@ -163,6 +141,44 @@ class QModelMenu(QMenu, _MenuMixin):
         for action in self.actions():
             if isinstance(action, QCommandRuleAction):
                 action._refresh()
+
+    def _disconnect(self) -> None:
+        self._app.menus.menus_changed.disconnect(self._on_registry_changed)
+
+    def _on_registry_changed(self, changed_ids: Set[str]) -> None:
+        if self._menu_id in changed_ids:
+            self.rebuild()
+
+    def update_from_context(
+        self, ctx: Mapping[str, object], _recurse: bool = True
+    ) -> None:
+        """Update the enabled/visible state of each menu item with `ctx`.
+
+        See `app_model.expressions` for details on expressions.
+
+        Parameters
+        ----------
+        ctx : Mapping
+            A namespace that will be used to `eval()` the `'enablement'` and
+            `'when'` expressions provided for each action in the menu.
+            *ALL variables used in these expressions must either be present in
+            the `ctx` dict, or be builtins*.
+        _recurse : bool
+            recursion check, internal use only
+        """
+        _update_from_context(self.actions(), ctx, _recurse=_recurse)
+
+    def rebuild(
+        self, include_submenus: bool = True, exclude: Optional[Collection[str]] = None
+    ) -> None:
+        """Rebuild menu by looking up self._menu_id in menu_registry."""
+        _rebuild(
+            menu=self,
+            app=self._app,
+            menu_id=self._menu_id,
+            include_submenus=include_submenus,
+            exclude=exclude,
+        )
 
 
 class QModelSubmenu(QModelMenu):
@@ -203,7 +219,7 @@ class QModelSubmenu(QModelMenu):
         # self.setVisible(expr.eval(ctx) if (expr := self._submenu.when) else True)
 
 
-class QModelToolBar(QToolBar, _MenuMixin):
+class QModelToolBar(QToolBar):
     """QToolBar that is built from a list of model menu ids."""
 
     def __init__(
@@ -217,18 +233,60 @@ class QModelToolBar(QToolBar, _MenuMixin):
     ) -> None:
         self._exclude = exclude
         QToolBar.__init__(self, parent)
-        _MenuMixin.__init__(self, menu_id, app)
+
+        # NOTE: code duplication with QModelMenu, but Qt mixins and multiple
+        # inheritance are problematic for some versions of Qt, and for typing
+        assert isinstance(menu_id, str), f"Expected str, got {type(menu_id)!r}"
+        self._menu_id = menu_id
+        self._app = Application.get_or_create(app) if isinstance(app, str) else app
+        self.setObjectName(menu_id)
+        self.rebuild()
+        self._app.menus.menus_changed.connect(self._on_registry_changed)
+        self.destroyed.connect(self._disconnect)
+
         if title is not None:
             self.setWindowTitle(title)
+
+    def addMenu(self, menu: QMenu) -> None:
+        """No-op for toolbar."""
+
+    def _disconnect(self) -> None:
+        self._app.menus.menus_changed.disconnect(self._on_registry_changed)
+
+    def _on_registry_changed(self, changed_ids: Set[str]) -> None:
+        if self._menu_id in changed_ids:
+            self.rebuild()
 
     def rebuild(
         self, include_submenus: bool = True, exclude: Optional[Collection[str]] = None
     ) -> None:
         """Rebuild toolbar by looking up self._menu_id in menu_registry."""
-        super().rebuild(include_submenus=include_submenus, exclude=self._exclude)
+        _rebuild(
+            menu=self,
+            app=self._app,
+            menu_id=self._menu_id,
+            include_submenus=include_submenus,
+            exclude=self._exclude,  # add comment as to why not use exclude?
+        )
 
-    def addMenu(self, menu: QMenu) -> None:
-        """No-op for toolbar."""
+    def update_from_context(
+        self, ctx: Mapping[str, object], _recurse: bool = True
+    ) -> None:
+        """Update the enabled/visible state of each menu item with `ctx`.
+
+        See `app_model.expressions` for details on expressions.
+
+        Parameters
+        ----------
+        ctx : Mapping
+            A namespace that will be used to `eval()` the `'enablement'` and
+            `'when'` expressions provided for each action in the menu.
+            *ALL variables used in these expressions must either be present in
+            the `ctx` dict, or be builtins*.
+        _recurse : bool
+            recursion check, internal use only
+        """
+        _update_from_context(self.actions(), ctx, _recurse=_recurse)
 
 
 class QModelMenuBar(QMenuBar):

--- a/src/app_model/backends/qt/_qmenu.py
+++ b/src/app_model/backends/qt/_qmenu.py
@@ -29,66 +29,6 @@ if TYPE_CHECKING:
     from qtpy.QtWidgets import QAction, QWidget
 
 
-def _rebuild(
-    menu: QMenu | QToolBar,
-    app: Application,
-    menu_id: str,
-    include_submenus: bool = True,
-    exclude: Optional[Collection[str]] = None,
-) -> None:
-    """Rebuild menu by looking up self._menu_id in menu_registry."""
-    menu.clear()
-    _exclude = exclude or set()
-
-    groups = list(app.menus.iter_menu_groups(menu_id))
-    n_groups = len(groups)
-    for n, group in enumerate(groups):
-        for item in group:
-            if isinstance(item, SubmenuItem) and include_submenus:
-                submenu = QModelSubmenu(item, app, parent=menu)
-                cast("QMenu", menu).addMenu(submenu)
-            elif item.command.id not in _exclude:  # type: ignore
-                action = QMenuItemAction(item, app=app, parent=menu)  # type: ignore
-                menu.addAction(action)
-        if n < n_groups - 1:
-            menu.addSeparator()
-
-
-def _update_from_context(
-    actions: Iterable[QAction], ctx: Mapping[str, object], _recurse: bool = True
-) -> None:
-    """Update the enabled/visible state of each menu item with `ctx`.
-
-    See `app_model.expressions` for details on expressions.
-
-    Parameters
-    ----------
-    actions : Iterable[QAction]
-        Actions to update.
-    ctx : Mapping
-        A namespace that will be used to `eval()` the `'enablement'` and
-        `'when'` expressions provided for each action in the menu.
-        *ALL variables used in these expressions must either be present in
-        the `ctx` dict, or be builtins*.
-    _recurse : bool
-        recursion check, internal use only
-    """
-    for action in actions:
-        if isinstance(action, QMenuItemAction):
-            action.update_from_context(ctx)
-        elif not QT6 and isinstance(menu := action.menu(), QModelMenu):
-            menu.update_from_context(ctx)
-        elif isinstance(parent := action.parent(), QModelMenu):
-            # FIXME: this is a hack for Qt6 that I don't entirely understand.
-            # QAction has lost the `.menu()` method, and it's a bit hard to find
-            # how to get to the parent menu now. Checking parent() seems to work,
-            # but I'm not sure if it's the right thing to do, and it leads to a
-            # recursion error.  I stop it with the _recurse flag here, but I wonder
-            # whether that will cause other problems.
-            if _recurse:
-                parent.update_from_context(ctx, _recurse=False)
-
-
 class QModelMenu(QMenu):
     """QMenu for a menu_id in an `app_model` MenusRegistry.
 
@@ -98,6 +38,8 @@ class QModelMenu(QMenu):
         Menu ID to look up in the registry.
     app : Union[str, Application]
         Application instance or name of application instance.
+    title : Optional[str]
+        Optional title for the menu, by default None
     parent : Optional[QWidget]
         Optional parent widget, by default None
     """
@@ -120,6 +62,7 @@ class QModelMenu(QMenu):
         self.rebuild()
         self._app.menus.menus_changed.connect(self._on_registry_changed)
         self.destroyed.connect(self._disconnect)
+        # ----------------------
 
         if title is not None:
             self.setTitle(title)
@@ -220,7 +163,21 @@ class QModelSubmenu(QModelMenu):
 
 
 class QModelToolBar(QToolBar):
-    """QToolBar that is built from a list of model menu ids."""
+    """QToolBar that is built from a list of model menu ids.
+
+    Parameters
+    ----------
+    menu_id : str
+        Menu ID to look up in the registry.
+    app : Union[str, Application]
+        Application instance or name of application instance.
+    exclude : Optional[Collection[str]]
+        Optional list of menu ids to exclude from the toolbar, by default None
+    title : Optional[str]
+        Optional title for the menu, by default None
+    parent : Optional[QWidget]
+        Optional parent widget, by default None
+    """
 
     def __init__(
         self,
@@ -243,6 +200,7 @@ class QModelToolBar(QToolBar):
         self.rebuild()
         self._app.menus.menus_changed.connect(self._on_registry_changed)
         self.destroyed.connect(self._disconnect)
+        # ----------------------
 
         if title is not None:
             self.setWindowTitle(title)
@@ -266,7 +224,7 @@ class QModelToolBar(QToolBar):
             app=self._app,
             menu_id=self._menu_id,
             include_submenus=include_submenus,
-            exclude=self._exclude,  # add comment as to why not use exclude?
+            exclude=self._exclude if exclude is None else exclude,
         )
 
     def update_from_context(
@@ -320,17 +278,65 @@ class QModelMenuBar(QMenuBar):
         _recurse : bool
             recursion check, internal use only
         """
-        for action in self.actions():
-            if isinstance(action, QMenuItemAction):
-                action.update_from_context(ctx)
-            elif not QT6 and isinstance(menu := action.menu(), QModelMenu):
-                menu.update_from_context(ctx)
-            elif isinstance(parent := action.parent(), QModelMenu):
-                # FIXME: this is a hack for Qt6 that I don't entirely understand.
-                # QAction has lost the `.menu()` method, and it's a bit hard to find
-                # how to get to the parent menu now. Checking parent() seems to work,
-                # but I'm not sure if it's the right thing to do, and it leads to a
-                # recursion error.  I stop it with the _recurse flag here, but I wonder
-                # whether that will cause other problems.
-                if _recurse:
-                    parent.update_from_context(ctx, _recurse=False)
+        _update_from_context(self.actions(), ctx, _recurse=_recurse)
+
+
+def _rebuild(
+    menu: QMenu | QToolBar,
+    app: Application,
+    menu_id: str,
+    include_submenus: bool = True,
+    exclude: Optional[Collection[str]] = None,
+) -> None:
+    """Rebuild menu by looking up `menu` in `Application`'s menu_registry."""
+    menu.clear()
+    _exclude = exclude or set()
+
+    groups = list(app.menus.iter_menu_groups(menu_id))
+    n_groups = len(groups)
+    for n, group in enumerate(groups):
+        for item in group:
+            if isinstance(item, SubmenuItem):
+                if include_submenus:
+                    submenu = QModelSubmenu(item, app, parent=menu)
+                    cast("QMenu", menu).addMenu(submenu)
+            elif item.command.id not in _exclude:
+                action = QMenuItemAction(item, app=app, parent=menu)
+                menu.addAction(action)
+        if n < n_groups - 1:
+            menu.addSeparator()
+
+
+def _update_from_context(
+    actions: Iterable[QAction], ctx: Mapping[str, object], _recurse: bool = True
+) -> None:
+    """Update the enabled/visible state of each menu item with `ctx`.
+
+    See `app_model.expressions` for details on expressions.
+
+    Parameters
+    ----------
+    actions : Iterable[QAction]
+        Actions to update.
+    ctx : Mapping
+        A namespace that will be used to `eval()` the `'enablement'` and
+        `'when'` expressions provided for each action in the menu.
+        *ALL variables used in these expressions must either be present in
+        the `ctx` dict, or be builtins*.
+    _recurse : bool
+        recursion check, internal use only
+    """
+    for action in actions:
+        if isinstance(action, QMenuItemAction):
+            action.update_from_context(ctx)
+        elif not QT6 and isinstance(menu := action.menu(), QModelMenu):
+            menu.update_from_context(ctx)
+        elif isinstance(parent := action.parent(), QModelMenu):
+            # FIXME: this is a hack for Qt6 that I don't entirely understand.
+            # QAction has lost the `.menu()` method, and it's a bit hard to find
+            # how to get to the parent menu now. Checking parent() seems to work,
+            # but I'm not sure if it's the right thing to do, and it leads to a
+            # recursion error.  I stop it with the _recurse flag here, but I wonder
+            # whether that will cause other problems.
+            if _recurse:
+                parent.update_from_context(ctx, _recurse=False)

--- a/src/app_model/backends/qt/_qmenu.py
+++ b/src/app_model/backends/qt/_qmenu.py
@@ -77,20 +77,7 @@ class QModelMenu(QMenu):
             Action ID to find. Note that `QCommandAction` have `ObjectName` set
             to their `command.id`
         """
-        return next((a for a in self.actions() if a.objectName() == object_name), None)
-
-    def _on_about_to_show(self) -> None:
-        # this would also be a reasonable place to call
-        for action in self.actions():
-            if isinstance(action, QCommandRuleAction):
-                action._refresh()
-
-    def _disconnect(self) -> None:
-        self._app.menus.menus_changed.disconnect(self._on_registry_changed)
-
-    def _on_registry_changed(self, changed_ids: Set[str]) -> None:
-        if self._menu_id in changed_ids:
-            self.rebuild()
+        return _find_action(self.actions(), object_name)
 
     def update_from_context(
         self, ctx: Mapping[str, object], _recurse: bool = True
@@ -122,6 +109,19 @@ class QModelMenu(QMenu):
             include_submenus=include_submenus,
             exclude=exclude,
         )
+
+    def _on_about_to_show(self) -> None:
+        # this would also be a reasonable place to call
+        for action in self.actions():
+            if isinstance(action, QCommandRuleAction):
+                action._refresh()
+
+    def _disconnect(self) -> None:
+        self._app.menus.menus_changed.disconnect(self._on_registry_changed)
+
+    def _on_registry_changed(self, changed_ids: Set[str]) -> None:
+        if self._menu_id in changed_ids:
+            self.rebuild()
 
 
 class QModelSubmenu(QModelMenu):
@@ -208,24 +208,16 @@ class QModelToolBar(QToolBar):
     def addMenu(self, menu: QMenu) -> None:
         """No-op for toolbar."""
 
-    def _disconnect(self) -> None:
-        self._app.menus.menus_changed.disconnect(self._on_registry_changed)
+    def findAction(self, object_name: str) -> Union[QAction, QModelMenu, None]:
+        """Find an action by its ObjectName.
 
-    def _on_registry_changed(self, changed_ids: Set[str]) -> None:
-        if self._menu_id in changed_ids:
-            self.rebuild()
-
-    def rebuild(
-        self, include_submenus: bool = True, exclude: Optional[Collection[str]] = None
-    ) -> None:
-        """Rebuild toolbar by looking up self._menu_id in menu_registry."""
-        _rebuild(
-            menu=self,
-            app=self._app,
-            menu_id=self._menu_id,
-            include_submenus=include_submenus,
-            exclude=self._exclude if exclude is None else exclude,
-        )
+        Parameters
+        ----------
+        object_name : str
+            Action ID to find. Note that `QCommandAction` have `ObjectName` set
+            to their `command.id`
+        """
+        return _find_action(self.actions(), object_name)
 
     def update_from_context(
         self, ctx: Mapping[str, object], _recurse: bool = True
@@ -245,6 +237,25 @@ class QModelToolBar(QToolBar):
             recursion check, internal use only
         """
         _update_from_context(self.actions(), ctx, _recurse=_recurse)
+
+    def rebuild(
+        self, include_submenus: bool = True, exclude: Optional[Collection[str]] = None
+    ) -> None:
+        """Rebuild toolbar by looking up self._menu_id in menu_registry."""
+        _rebuild(
+            menu=self,
+            app=self._app,
+            menu_id=self._menu_id,
+            include_submenus=include_submenus,
+            exclude=self._exclude if exclude is None else exclude,
+        )
+
+    def _disconnect(self) -> None:
+        self._app.menus.menus_changed.disconnect(self._on_registry_changed)
+
+    def _on_registry_changed(self, changed_ids: Set[str]) -> None:
+        if self._menu_id in changed_ids:
+            self.rebuild()
 
 
 class QModelMenuBar(QMenuBar):
@@ -340,3 +351,7 @@ def _update_from_context(
             # whether that will cause other problems.
             if _recurse:
                 parent.update_from_context(ctx, _recurse=False)
+
+
+def _find_action(actions: Iterable[QAction], object_name: str) -> Union[QAction, None]:
+    return next((a for a in actions if a.objectName() == object_name), None)

--- a/tests/test_qt/test_qmenu.py
+++ b/tests/test_qt/test_qmenu.py
@@ -111,7 +111,7 @@ def test_shortcuts(qtbot: "QtBot", full_app: "FullApp") -> None:
 
     copy_action = menu.findAction(app.Commands.COPY)
 
-    with qtbot.waitSignal(copy_action.triggered, timeout=1000):
+    with qtbot.waitSignal(copy_action.triggered, timeout=2000):
         qtbot.keyClicks(win, "C", Qt.KeyboardModifier.ControlModifier)
 
     paste_action = menu.findAction(app.Commands.PASTE)

--- a/tests/test_qt/test_qmenu.py
+++ b/tests/test_qt/test_qmenu.py
@@ -1,11 +1,13 @@
+from __future__ import annotations
+
 import sys
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, cast
 
 import pytest
 from qtpy.QtCore import Qt
 from qtpy.QtWidgets import QAction, QMainWindow
 
-from app_model.backends.qt import QModelMenu
+from app_model.backends.qt import QModelMenu, QModelToolBar
 
 if TYPE_CHECKING:
     from pytestqt.plugin import QtBot
@@ -16,10 +18,13 @@ SEP = ""
 LINUX = sys.platform.startswith("linux")
 
 
-def test_menu(qtbot: "QtBot", full_app: "FullApp") -> None:
+@pytest.mark.parametrize("MenuCls", [QModelMenu, QModelToolBar])
+def test_menu(
+    MenuCls: type[QModelMenu] | type[QModelToolBar], qtbot: QtBot, full_app: FullApp
+) -> None:
     app = full_app
 
-    menu = QModelMenu(app.Menus.EDIT, app)
+    menu = MenuCls(app.Menus.EDIT, app)
     qtbot.addWidget(menu)
 
     # The "" are separators, according to our group settings in full_app
@@ -28,12 +33,12 @@ def test_menu(qtbot: "QtBot", full_app: "FullApp") -> None:
 
     # check that triggering the actions calls the associated commands
     for cmd in (app.Commands.UNDO, app.Commands.REDO):
-        action: QAction = menu.findAction(cmd)
+        action = cast(QAction, menu.findAction(cmd))
         with qtbot.waitSignal(action.triggered):
             action.trigger()
             getattr(app.mocks, cmd).assert_called_once()
 
-    redo_action: QAction = menu.findAction(app.Commands.REDO)
+    redo_action = cast(QAction, menu.findAction(app.Commands.REDO))
 
     assert redo_action.isVisible()
     assert redo_action.isEnabled()
@@ -59,8 +64,10 @@ def test_menu(qtbot: "QtBot", full_app: "FullApp") -> None:
     with pytest.raises(NameError, match="Names required to eval this expression"):
         menu.update_from_context({})
 
+    menu._disconnect()
 
-def test_submenu(qtbot: "QtBot", full_app: "FullApp") -> None:
+
+def test_submenu(qtbot: QtBot, full_app: FullApp) -> None:
     app = full_app
 
     menu = QModelMenu(app.Menus.FILE, app)
@@ -97,7 +104,7 @@ def test_submenu(qtbot: "QtBot", full_app: "FullApp") -> None:
 
 @pytest.mark.filterwarnings("ignore:QPixmapCache.find:")
 @pytest.mark.skipif(LINUX, reason="Linux keytest not working")
-def test_shortcuts(qtbot: "QtBot", full_app: "FullApp") -> None:
+def test_shortcuts(qtbot: QtBot, full_app: FullApp) -> None:
     app = full_app
 
     win = QMainWindow()
@@ -119,7 +126,7 @@ def test_shortcuts(qtbot: "QtBot", full_app: "FullApp") -> None:
         qtbot.keyClicks(win, "V", Qt.KeyboardModifier.ControlModifier)
 
 
-def test_toggled_menu_item(qtbot: "QtBot", full_app: "FullApp") -> None:
+def test_toggled_menu_item(qtbot: QtBot, full_app: FullApp) -> None:
     app = full_app
     menu = QModelMenu(app.Menus.HELP, app)
     qtbot.addWidget(menu)
@@ -130,3 +137,29 @@ def test_toggled_menu_item(qtbot: "QtBot", full_app: "FullApp") -> None:
 
     menu.update_from_context({"thing_toggled": False})
     assert not action.isChecked()
+
+
+@pytest.mark.parametrize("MenuCls", [QModelMenu, QModelToolBar])
+def test_menu_events(
+    MenuCls: type[QModelMenu] | type[QModelToolBar], qtbot: QtBot, full_app: FullApp
+) -> None:
+    app = full_app
+    menu = MenuCls(app.Menus.EDIT, app)
+
+    qtbot.addWidget(menu)
+
+    # The "" are separators, according to our group settings in full_app
+    menu_texts = [a.text() for a in menu.actions()]
+    assert menu_texts == ["AtTop", SEP, "Undo", "Redo", SEP, "Copy", "Paste"]
+
+    # simulate something changing the edit menu... normally this would be
+    # triggered by a dispose() call, but that's a bit hard to do currently with the
+    # test app fixture.
+    copy_item = next(
+        x for x in full_app.menus._menu_items["edit"] if x.command.title == "Copy"
+    )
+    full_app.menus._menu_items["edit"].pop(copy_item)
+    full_app.menus.menus_changed.emit(app.Menus.EDIT)
+
+    menu_texts = [a.text() for a in menu.actions()]
+    assert menu_texts == ["AtTop", SEP, "Undo", "Redo", SEP, "Paste"]

